Added performance impact section to the docs (#279)
elastic/apm-aws-lambda: docs/monitoring-aws-lambda.asciidoc

By using an AWS Lambda extension, Elastic APM agents can send data to a local Lambda extension process, and that process will forward data on to APM Server asynchronously. The Lambda extension ensures that any potential latency between the Lambda function and the APM Server instance will not cause latency in the request flow of the Lambda function itself.

[[aws-lambda-overhead]]
== Performance impact and overhead

> As described in the section on <<aws-lambda-arch>>, using Elastic APM with AWS Lambda requires to include the Elastic APM AWS Lambda extension and a corresponding Elastic APM agent to the Lambda runtime.

Reviewer: ```suggestion
As described in the section on <<aws-lambda-arch>>, using Elastic APM with AWS Lambda requires one to include the Elastic APM AWS Lambda extension and a corresponding Elastic APM agent with the Lambda function.
```

Grammar. Totally feel free to ignore these suggestions. :)

Reviewer 2: I'll add to Trent's suggestions with this:

```suggestion
As described in <<aws-lambda-arch>>, using Elastic APM with AWS Lambda requires adding both the Elastic APM AWS Lambda extension and a corresponding Elastic APM agent to the Lambda runtime.
```

These components may introduce a small overhead on the size of your function's deployment package as well as the execution duration of your function's invocations. 

=== Impact on the deployment package size

> These compotents contribute a little to the uncompressed deployment package size of your Lambda function. The Elastic APM AWS Lambda extension and each of the APM agents have an uncompressed package size that is below 10MB. Thus, the overall impact of using Elastic APM on the uncompressed deployment package size of your Lambda function less than 20MB.

Reviewer: ```suggestion
These components contribute a little to the uncompressed deployment package size of your Lambda function. The Elastic APM AWS Lambda extension and each of the APM agents have an uncompressed package size that is below 10MB. Thus, the overall impact of using Elastic APM on the uncompressed deployment package size of your Lambda function is less than 20MB.
```

FWIW the node.js APM agent layer is *17MB* uncompressed:

```
% ./dev-utils/make-distribution.sh
...
Created build/dist/elastic-apm-node-lambda-layer.zip
...
% du -sh build/dist/nodejs
 17M	build/dist/nodejs
```

A dev build of the extension is 10MB for me.

Reviewer: Note that if we made it a goal, I'm confident we could get that down under ~10MB pretty easily.

=== Performance impact

> The <<aws-lambda-arch>> comes with the adavantage that APM data dispatching is decoupled from your function's request processing. 

Reviewer 2: I'm not sure this link is necessary. This reads as:

> The [APM Architecture for AWS Lambda](https://apm-aws-lambda_279.docs-preview.app.elstc.co/guide/en/apm/guide/master/aws-lambda-arch.html) comes with the adavantage that APM data dispatching is decoupled from your function’s request processing.

Are you meaning to use the extension name instead?

 ```suggestion
 An advantage of the Elastic APM AWS Lambda extension is that APM data dispatching is decoupled from your function's request processing.
```

Reviewer 2: Oh, I think I get it now. The advantage is because of the architecture? I still don't think we need to link to that page again, but maybe these words are better:

```suggestion
An advantage of the Elastic APM AWS Lambda extension architecture is that APM data dispatching is decoupled from your function's request processing.
```

> The Elastic APM AWS Lambda extension flushes APM data to the Elastic backend _after_ your function reponds to the client's request. Thus, it does not affect the latency of the client's request. 

Reviewer: ```suggestion
The Elastic APM AWS Lambda extension flushes APM data to the Elastic backend _after_ your function responds to the client's request. Thus, it does not affect the latency of the client's request. 
```

typo

However, the extension's flushing of APM data contributes to the overall execution time of the function invocation. 
The <<aws-lambda-config-data-forwarder-timeout>> config option with the related _exponential backoff algorithm_ limits and allows to control the impact the extension may have on the function's overall execution time. 

When your function experiences a cold start, the Elastic APM AWS Lambda extension needs to be initialized and, thus, slightly increases the cold start duration (in the range of tens of milliseconds) of your function. 

> APM agents enrich your application's code with measurement code that collects APM data. This measurement code introduces a small performance overhead to your application, which is usually in a negligible range. This matter of fact is no different with lambda functions. The concrete performance overhead introduced by APM agents highly depends on the configuration of the agent and on the characteristics of your function's code. The following agent-specific documentation pages provide insights and instructions on tuning the performance the APM agents:

Reviewer: ```suggestion
APM agents enrich your application's code with measurement code that collects APM data. This measurement code introduces a small performance overhead to your application, which is usually in a negligible range. The same is true with Lambda functions. The concrete performance overhead introduced by APM agents highly depends on the configuration of the agent and on the characteristics of your function's code. The following agent-specific documentation pages provide insights and instructions on tuning the performance the APM agents:
```
